src/FiniteVolumeMLMC.hpp

```diff
@@ -89,13 +103,39 @@ class FiniteVolumeMLMC : public Upscale
 
     void MakeCoarseSolver();
 
+    void SetEssentialData(const mfem::Vector& new_data,
+                          int special_vertex_dofs);
+
+    /// hack, for solve with essential vertex boundary conditions
+    void SolveFineEssU(const mfem::BlockVector& x, mfem::BlockVector& y) const;
+    /// hack
+    void SolveEssU(int level, const mfem::BlockVector& x, mfem::BlockVector& y) const;
+
 private:
     const mfem::Vector& weight_;
     const mfem::HypreParMatrix& edge_d_td_;
     const mfem::SparseMatrix& edge_boundary_att_;
     const mfem::Array<int>& ess_attr_;
 
     const UpscaleParameters& param_;
+
+    const mfem::Array<int>& ess_u_marker_;
+    mfem::Vector ess_u_data_;
+    mfem::Array<int> coarse_ess_u_marker_;
+    mfem::Vector coarse_ess_u_data_;
+    /// @todo not all of the next four bools are necessary?
+    bool impose_ess_u_conditions_;
+    bool ess_u_matrix_eliminated_;
+    bool coarse_impose_ess_u_conditions_;
+    bool coarse_ess_u_matrix_eliminated_;
+
+    std::unique_ptr<mfem::BlockVector> ess_u_finerhs_correction_;
+    std::unique_ptr<mfem::BlockVector> ess_u_coarserhs_correction_;
+
+    std::unique_ptr<mfem::SparseMatrix> fine_DTelim_trace_;
+    std::unique_ptr<mfem::SparseMatrix> coarse_DTelim_trace_;
+
+    void CoarsenEssentialVertexBoundary(int special_vertex_dofs);
```

Comment:
Add `level` argument, probably other modifications.

Also *document* because even though I wrote it I don't know exactly what this does (!).